flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/LocalInputPreferredSlotSharingStrategy.java

```diff
@@ -0,0 +1,297 @@
+
+	LocalInputPreferredSlotSharingStrategy(
+			final SchedulingTopology topology,
+			final Set<SlotSharingGroup> logicalSlotSharingGroups,
+			final Set<CoLocationGroupDesc> coLocationGroups) {
+
+		this.executionSlotSharingGroupMap = new ExecutionSlotSharingGroupBuilder(
+			topology,
+			logicalSlotSharingGroups,
+			coLocationGroups).build();
+	}
+
+	@Override
+	public ExecutionSlotSharingGroup getExecutionSlotSharingGroup(final ExecutionVertexID executionVertexId) {
+		return executionSlotSharingGroupMap.get(executionVertexId);
+	}
+
+	@Override
+	public Set<ExecutionSlotSharingGroup> getExecutionSlotSharingGroups() {
+		return new HashSet<>(executionSlotSharingGroupMap.values());
+	}
+
+	static class Factory implements SlotSharingStrategy.Factory {
+
+		public LocalInputPreferredSlotSharingStrategy create(
+				final SchedulingTopology topology,
+				final Set<SlotSharingGroup> logicalSlotSharingGroups,
+				final Set<CoLocationGroupDesc> coLocationGroups) {
+
+			return new LocalInputPreferredSlotSharingStrategy(topology, logicalSlotSharingGroups, coLocationGroups);
+		}
+	}
+
+	private static class ExecutionSlotSharingGroupBuilder {
+		private final SchedulingTopology topology;
+
+		private final Map<JobVertexID, SlotSharingGroupId> slotSharingGroupMap;
+
+		private final Map<JobVertexID, CoLocationGroupDesc> coLocationGroupMap;
+
+		private final Map<ExecutionVertexID, ExecutionSlotSharingGroup> executionSlotSharingGroupMap;
+
+		final Map<CoLocationConstraintDesc, ExecutionSlotSharingGroup> constraintToExecutionSlotSharingGroupMap;
+
+		final Map<SlotSharingGroupId, List<ExecutionSlotSharingGroup>> executionSlotSharingGroups;
+
+		private final Map<ExecutionSlotSharingGroup, Set<JobVertexID>> assignedJobVerticesForGroups;
+
+		private ExecutionSlotSharingGroupBuilder(
+				final SchedulingTopology topology,
+				final Set<SlotSharingGroup> logicalSlotSharingGroups,
+				final Set<CoLocationGroupDesc> coLocationGroups) {
+
+			this.topology = checkNotNull(topology);
+
+			this.slotSharingGroupMap = new HashMap<>();
+			for (SlotSharingGroup slotSharingGroup : logicalSlotSharingGroups) {
+				for (JobVertexID jobVertexId : slotSharingGroup.getJobVertexIds()) {
+					slotSharingGroupMap.put(jobVertexId, slotSharingGroup.getSlotSharingGroupId());
+				}
+			}
+
+			this.coLocationGroupMap = new HashMap<>();
+			for (CoLocationGroupDesc coLocationGroup : coLocationGroups) {
+				for (JobVertexID jobVertexId : coLocationGroup.getVertices()) {
+					coLocationGroupMap.put(jobVertexId, coLocationGroup);
+				}
+			}
+
+			executionSlotSharingGroupMap = new HashMap<>();
+			constraintToExecutionSlotSharingGroupMap = new HashMap<>();
+			executionSlotSharingGroups = new HashMap<>();
+			assignedJobVerticesForGroups = new IdentityHashMap<>();
+		}
+
+		/**
+		 * Build ExecutionSlotSharingGroups for all vertices in the topology.
+		 * The ExecutionSlotSharingGroup of a vertex is determined in order below:
```

Comment:
One known defect is the grouping of execution vertices from brother job vertices are possible suboptimal because it does not take connections to consumers vertices into account.

Example: A(parallelism=4) --forward--> C(parallelism=4), B(parallelism=2) --rescale--> C
Execution edges are: A1->C1, A2->C2, A3->C3, A4->C4; B1->C1, B1->C2, B2->C3,B2->C4
Optimal grouping: {A1,B1,C1}{A2,C2}{A3,B2,C3}{A4,C4}     ->   there would be 2 remote edges
Current grouping result: {A1,B1,C1} {A2,B2,C2}  {A3,C3}  {A4,C4}     ->   there would be 3 remote edges

Reply:
If we do not want to do the full complicated connected graph analysis (harder to maintain), we could a lighter optimisation now or later. After the suggested traversal, we could do one more traversal where we try to move producers closer to their consumers if the producers are not already bounded by other co-located consumers or possibly by a co-location constraint. Parent producers of the producers (if parent ones have less parallelism) should not bound them for the move if the second traversal is bottom->up.

Reply:
Result of our offline discussion with @tillrohrmann:
The described suboptimal case is not a strict regression because the existing `SlotSharingManager` can already produce this suboptimal case. The `SlotSharingStrategy` is already designed to be pluggable. Therefore, we can focus on the overall new bulk allocation implementation and optimize the simplest `SlotSharingStrategy` approach, suggested in this PR, later.